java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java

```diff
@@ -211,7 +211,8 @@ public PutObserver(VectorUnloader unloader, ClientCallStreamObserver<ArrowMessag
     @Override
     public void putNext() {
       ArrowRecordBatch batch = unloader.getRecordBatch();
-      while (!observer.isReady()) {
+      // Check the futureResult in case server sent an exception
+      while (!observer.isReady() && !futureResult.isDone()) {
```

Comment:
I'm curious: is it customary to do such busy wait in Java?

Reply:
There is a callback available in this case, though it would complicate the code quite a bit. Let me take a stab at it...